[LT][104029130] Support additional pin query params

luketower opened the pull request:
[Pivotal Story](https://www.pivotaltracker.com/story/show/104029130)

In order to match the pin result with the new search filter
display values, it is necessary to send more data to the
map/pin route.

c0 commented on app/coffeescript/components/mixins/map_utils.coffee:
```diff
@@ -3,12 +3,13 @@ define [ 'jquery' ], ( $ ) ->
+      assetHostSelector:      'meta[name="asset_host"]'
+      spinnerSelector:        '.spinner'
+      refinementsSelector:    '.pageInfo[name="refinements"]'
+      pinRefinementsSelector: '.pageInfo[name="pin_search_refinements"]'
```
Does this need to be specific to just pins? Can we consolidate into one meta tag?

luketower replied:
I thought it was a good idea to separate these as the refinements were being used in other places in the application.

c0 replied:
Do you have an example of both? I'm interested in the differences.

sbrauer replied:
My preference would be to reuse the existing refinements meta tag if possible (I can't think of any reason it wouldn't be).

mpelzsherman replied:
Here are some examples:

URL: http://local.apartmentguide.com/map?state=Georgia&city=Atlanta&min_price=1300&max_price=1500

``` javascript
$('.pageInfo[name="refinements"]')
[]
$('.pageInfo[name="pin_search_refinements"]')
[<meta class=​"pageInfo" content=​"max_price=1500&min_price=1300" name=​"pin_search_refinements">​]
```

This is the problem use case. We can't just use the existing `refinements` element from the DOM because we need the min & max price data to pass to the info_window.

http://local.apartmentguide.com/map?state=Georgia&city=Atlanta&refinements=2-beds-1z141y8&min_price=1300&max_price=1500

``` javascript
$('.pageInfo[name="pin_search_refinements"]')
[<meta class=​"pageInfo" content=​"refinements=2-beds-1z141y8&max_price=1500&min_price=1300" name=​"pin_search_refinements">​]
$('.pageInfo[name="refinements"]')
[<meta class=​"pageInfo" content=​"2-beds-1z141y8" name=​"refinements">​]
```

Note that the `refinements` element does not include the "refinements=" string, because map.js is going to add it.

Hope this is clear. Basically we don't want to try to make the `refinements` tag do double-duty because of all the dependencies from both AG and Rent. Happy to discuss further.

luketower committed:
[LT][104029130] Read pin query params from new head tag on AG

In order to match the pin result with the new search filter
display values, it is necessary to send more data to the
map/pin route. This data is available in a new head tag on
AG.

mpelzsherman commented:
Per conversation with @c0, we're going to try a different approach - passing the data into info_window.js when it's attached, instead of grabbing from the DOM. Also we're going to try using the `ApartmentGuide' config object, which already has the data we need.

mpelzsherman committed:
[MPS/LT/CW][104029130] refactor method of passing pin refinements into info_window component

 This approach was suggested by Caleb W - passing the refinements directly when we attach the component rather than grabbing from the meta tag.

c0 commented on app/coffeescript/components/data/info_window.coffee:
```diff
@@ -3,22 +3,20 @@
     @defaultAttrs
       pinRoute: "/map/pin/"
-
+      pinRefinements: ''
```
Why is this a string rather than a key/value pair?

c0 replied:
We can also remove the `pin` prefix. It avoids the Smurf Naming Convention ;)

![screenshot 2015-12-07 12 12 46](https://cloud.githubusercontent.com/assets/338738/11633922/dca4a876-9cdb-11e5-8135-2feaaeb59da4.png)

mpelzsherman committed:
[MPS/LT/CW][104029130] removed superfluous "pin" prefix

mpelzsherman commented on dist/components/data/info_window.js:
```diff
@@ -1,13 +1,14 @@
     this.defaultAttrs({
-      pinRoute: "/map/pin/"
+      route: "/map/pin/",
+      refinements: ''
```
@c0 - this is being used "as is" to form the query string, because each app has different requirements around this.

Should we change variable name to "queryString"?

c0 replied:
A query string makes it difficult for the app to implement. It's only used to make the post request, which is an internal part of the info window component.

luketower committed:
[LT][104029130] Build query params from refinements hash

mpelzsherman committed:
[MPS/LT][104029130] Added a test to info_window_spec for multiple refinements.

mpelzsherman commented on test/spec/components/data/info_window_spec.coffee:
```diff
@@ -3,20 +3,65 @@ define ['jquery'], ($) ->
+                            dim_id: '123xyz')
+        expect(@component.queryParams()).toEqual('?refinements=2-beds-123xyz')
+
+      it 'creates a query string with beds and baths', ->
```
@luketower - I added this failing test - I think it needs to pass.

mpelzsherman committed:
[MPS][104029130] Added .jshintrc to avoid for-in warnings from CodeClimate.

mpelzsherman committed:
[MPS][104029130] Backing out .jshintrc file in favor of excluding "dist" dir.

mpelzsherman committed:
[MPS][104029130] Added .codeclimate.yml

mpelzsherman committed:
Revert "[MPS][104029130] Added .codeclimate.yml"

This reverts commit 24fb9db7e0845190ba9cb03dfee2f229f150b23d.

mpelzsherman committed:
[MPS][104029130] Use "for own" to avoid potential iteration problems.

http://stackoverflow.com/questions/1963102/what-does-the-jslint-error-body-of-a-for-in-should-be-wrapped-in-an-if-statemen

mpelzsherman committed:
[MPS][104029130] Add .idea to .gitignore (for RubyMine users)

mpelzsherman committed:
[MPS][104029130] Enable eslint engine on JS files, with custom rules in .eslintrc.yml.

luketower committed:
[LT][104029130] Handle multiple refinements when creating query string

luketower committed:
Merge branch 'lt_104029130_beds_and_price_filtering' of https://github.com/rentpath/map.js into lt_104029130_beds_and_price_filtering

Conflicts:
  app/coffeescript/components/data/info_window.coffee
  dist/components/data/info_window.js

mpelzsherman committed:
[MPS][104029130] Try using airbnb eslint style guide.

mpelzsherman commented on app/coffeescript/components/data/info_window.coffee:
```diff
@@ -3,22 +3,42 @@
+    @queryParams = ->
+      return "" if _.isEmpty(@attr.refinements)
+      results = {names: [], ids: [], filters: []}
+      for name, obj of @attr.refinements
```
We should use 

``` coffeescript
  for **own** name, obj
```

here. See http://stackoverflow.com/questions/1963102/what-does-the-jslint-error-body-of-a-for-in-should-be-wrapped-in-an-if-statemen.

mpelzsherman committed:
[MPS][104029130] Use "for own" to protect against iteration scope

c0 commented on .gitignore:
```diff
@@ -4,3 +4,4 @@ node_modules/
 npm-debug.log
 npm-debug.log.*
 app/js
+.idea/
```
:+1: We don't want to give git any ideas.

colinrymer commented:
Why are the `dist` dir contents included?

c0 commented on app/coffeescript/components/data/info_window.coffee:
```diff
@@ -3,22 +3,42 @@
+        else
+          results.names.push obj.dim_name
+          results.ids.push obj.dim_id
+      "?#{@formatRefinements(results)}#{@formatFilters(results.filters, results.names)}"
```
Can we separate the logic from building the url string? For example, use this function to build a generic `results` and other to take an object and turn it into a query string.

luketower committed:
[LT/MPS][104029130] Break up logic for queryParams

mpelzsherman commented:
@colinrymer - we are including JS `dist` contents in all of our other bower components.

If we want to change this, we should be consistent and do it in separate PR's.

turbodog99 commented:
If we want to change the way we're distributing JS, that's a chore for another story. Merging.

:+1:
